diffrax/_term.py

```diff
@@ -776,3 +785,139 @@ def _to_vjp(_y, _diff_args, _diff_term):
         dy, vjp = jax.vjp(_to_vjp, y, diff_args, diff_term)
         da_y, da_diff_args, da_diff_term = vjp((-(a_y**ω)).ω)
         return dy, da_y, da_diff_args, da_diff_term
+
+
+# The Underdamped Langevin SDE trajectory consists of two components: the position
+# `x` and the velocity `v`. Both of these have the same shape. So, by LangevinX we
+# denote the shape of the x component, and by LangevinTuple we denote the shape of
+# the tuple (x, v).
+LangevinX = PyTree[Shaped[Array, "?*langevin"], "LangevinX"]
+LangevinTuple = tuple[LangevinX, LangevinX]
+
+
+class LangevinDiffusionTerm(
+    AbstractTerm[LangevinX, Union[LangevinX, AbstractBrownianIncrement]]
+):
+    r"""Represents the diffusion term in the Langevin SDE:
+
+    $d \mathbf{x}_t = \mathbf{v}_t dt$
+
+    $d \mathbf{v}_t = - \gamma \mathbf{v}_t dt - u
+    \nabla f( \mathbf{x}_t ) dt + \sqrt{2 \gamma u} d W_t.$
+    """
+
+    gamma: LangevinX
+    u: LangevinX
+    control: AbstractBrownianPath
+
+    def vf(self, t: RealScalarLike, y: LangevinTuple, args: Args) -> LangevinX:
+        x, v = y
+
+        def _fun(_gamma, _u, _v):
+            return jnp.sqrt(2 * _gamma * _u) * jnp.ones(_v.shape, _v.dtype)
+
+        d_v = jtu.tree_map(_fun, self.gamma, self.u, v)
+        return d_v
+
+    def contr(
+        self, t0: RealScalarLike, t1: RealScalarLike, **kwargs
+    ) -> Union[LangevinX, AbstractBrownianIncrement]:
+        return self.control.evaluate(t0, t1, **kwargs)
+
+    def prod(self, vf: LangevinX, control: LangevinX) -> LangevinTuple:
+        dv = vf
+        dw = control
+        v_out = jtu.tree_map(operator.mul, dv, dw)
+        x_out = jtu.tree_map(jnp.zeros_like, v_out)
+        return x_out, v_out
+
+
+def _broadcast_pytree(source, target_tree_shape):
+    # Requires that source is a prefix tree of target_tree_shape
+    def _inner_broadcast(_src_arr, _inner_tree_shape):
+        _arr = jnp.asarray(_src_arr)
+        return jtu.tree_map(
+            lambda _leaf: jnp.broadcast_to(_arr, _leaf.shape), _inner_tree_shape
+        )
+
+    return jtu.tree_map(_inner_broadcast, source, target_tree_shape)
+
+
+class LangevinDriftTerm(AbstractTerm):
+    gamma: LangevinX
+    u: LangevinX
+    grad_f: Callable[[LangevinX], LangevinX]
+
+    def vf(self, t: RealScalarLike, y: LangevinTuple, args: Args) -> LangevinTuple:
+        x, v = y
+        f_x = self.grad_f(x)
+        d_x = v
+        d_v = jtu.tree_map(
+            lambda _gamma, _u, _v, _f_x: -_gamma * _v - _u * _f_x,
+            self.gamma,
+            self.u,
+            v,
+            f_x,
+        )
+        d_y = (d_x, d_v)
+        return d_y
+
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> RealScalarLike:
+        return t1 - t0
+
+    def prod(self, vf: LangevinTuple, control: RealScalarLike) -> LangevinTuple:
+        return jtu.tree_map(lambda _vf: control * _vf, vf)
+
+
+def make_langevin_term(
+    gamma: PyTree[ArrayLike],
+    u: PyTree[ArrayLike],
```

Comment:
Should these by pytrees with the same structure?

Reply:
Originally yes, but given that you told me to go with the first option (regarding broadcasting these to the structure and shape of `x0`), this is no longer the case, so they needn't have the same structure. In fact a common use-case is that `u` is a scalar, and `gamma` has the same shape&structure as `x0`.

Reply:
Got it!